File: src/models/exchangerate.nim
import std/[times, options]
import orm/orm
import currency, user

type
  ExchangeRate* = object of Model
    rate*: float
    reverse_rate*: float
    valid_date*: DateTime
    rate_source*: string
    bnb_rate_id*: string
    is_active*: bool
    notes*: string
    from_currency_id*: int64
    to_currency_id*: int64
    created_by_id*: Option[int64]
    created_at*: DateTime
    updated_at*: DateTime

proc newExchangeRate*(
  rate = 1.0,
  reverse_rate = 1.0,
  valid_date = now(),
  rate_source = "MANUAL",
  from_currency_id: int64 = 0,
  to_currency_id: int64 = 0
): ExchangeRate =
  ExchangeRate(
    id: 0,
    rate: rate,
    reverse_rate: reverse_rate,
    valid_date: valid_date,
    rate_source: rate_source,
    bnb_rate_id: "",
    is_active: true,
    notes: "",
    from_currency_id: from_currency_id,
    to_currency_id: to_currency_id,
    created_by_id: none(int64),
    created_at: now(),
    updated_at: now()
  )

File: baraba_ub/apps/baraba_umbrella/lib/baraba_umbrella/accounting/exchange_rate.ex
defmodule BarabaUmbrella.Accounting.ExchangeRate do
  use Ecto.Schema
  import Ecto.Changeset

  @primary_key {:id, :binary_id, autogenerate: true}
  @foreign_key_type :binary_id

  schema "exchange_rates" do
    field :rate, :decimal
    field :reverse_rate, :decimal
    field :valid_date, :date
    field :rate_source, :string, default: "MANUAL"
    field :bnb_rate_id, :string
    field :is_active, :boolean, default: true
    field :notes, :string
    
    belongs_to :from_currency, BarabaUmbrella.Accounting.Currency
    belongs_to :to_currency, BarabaUmbrella.Accounting.Currency
    belongs_to :created_by, BarabaUmbrella.Accounting.User

    timestamps()
  end

  @doc false
  def changeset(exchange_rate, attrs) do
    exchange_rate
    |> cast(attrs, [:rate, :reverse_rate, :valid_date, :rate_source, :bnb_rate_id, :is_active, :notes, :from_currency_id, :to_currency_id, :created_by_id])
    |> validate_required([:rate, :valid_date, :from_currency_id, :to_currency_id])
  end
end
